src/ch07/c1_breed_rats.py

```diff
@@ -0,0 +1,463 @@
+"""Efficiently breed rats to an average weight of 50000 grams.
+
+Use genetic algorithm on a mixed population of male and female rats.
+
+"""
+import time
+import random
+import statistics
+
+
+class BreedRats:
+    """Efficiently breed rats to an average weight of **target_wt**.
+
+    Use genetic algorithm on a mixed population of male and female rats.
+
+    Weights and number of each gender vary and can be set by modifying the
+    following:
+
+    Args:
+        num_males (int): Number of male rats in population.
+            Default is ``4``.
+        num_females (int): Number of female rats in population.
+            Default is ``16``.
+        target_wt (int): Target weight in grams. Default is ``50000``.
+        gen_limit (int): Generational cutoff to stop breeding program.
+            Default is ``500``.
+
+    """
+
+    # pylint: disable=too-many-instance-attributes,too-many-public-methods
+    # Limit is for instance attributes and public methods are 7 and 20,
+    # but analysis like this requires many constants.
+    # I am opting to make them modifiable in something that is isn't a
+    # dictionary.
+    # If there is a better way, please submit an issue for discussion.
+
+    def __init__(self, num_males: int = 4, num_females: int = 16,
+                 target_wt: int = 50000, gen_limit: int = 500):
+        """Initialize class."""
+        self._min_wt = 200
+        self._max_wt = 600
+        self._male_mode_wt = 300
+        self._female_mode_wt = 250
+        self._mut_odds = 0.01
+        self._mut_min = 0.5
+        self._mut_max = 1.2
+        self._litter_sz = 8
+        self._litters_per_yr = 10
+
+        self._num_males = num_males
+        self._num_females = num_females
+        self._target_wt = target_wt
+        self._gen_limit = gen_limit
+
+    @property
+    def min_wt(self):
+        """int: Minimum weight of adult rat in initial population.
+
+        Default is ``200``.
+        """
+        return self._min_wt
+
+    @min_wt.setter
+    def min_wt(self, value: int):
+        self._min_wt = value
```

Comment:
I considered making the properties mixedCase, like I've seen in other packages, but PEP8 advises to follow function naming rules for [method names](https://www.python.org/dev/peps/pep-0008/#method-names-and-instance-variables). Furthermore, PEP8 advises [function names](https://www.python.org/dev/peps/pep-0008/#function-and-variable-names) should be lowercase with underscores.

Technically, ``minWt `` has one fewer key press than ``min_wt``. This will keep me up at night.